### DISCUSSION TITLE
How to correctly format language settings in an extension file?

### DISCUSSION THREAD

**Discussion by pftnhr** [datenstrom/community#898] [Ask a question] [answered]
For the maillink extension I would like to offer more languages, for this I have used the following code:

```php
// Handle initialisation
public function onLoad($yellow) {
$this->yellow = $yellow;
$this->yellow->system->setDefault("MailAddress", "");
$this->yellow->language->setDefaults(array(
	"Language: en",
	"MailLinktext: Drop me a line",
	"Language: de",
	"MailLinktext: Schreib mir ein paar Zeilen",
	"Language: fr",
	"MailLinktext: Contactez-moi",
	"Language: it",
	"MailLinktext: Scrivimi una riga"));
}
```
Unfortunately, none of the strings are written into the yellow-language.ini. I have tested it several times in a fresh, current installation and only `$this->yellow->system->setDefault("MailAddress", "");` is written to yellow-system.ini.

Test environment:

* current Yellow version
* clean installation
* no further extensions

What else have I tried?

* I installed Giovanni's [`yellow-link`](https://github.com/GiovanniSalmeri/yellow-link) in my sandbox to see if it was Yellow. It wasn't. The language.ini got updated.
* I deleted the `status: experimental` from extension.ini but that wasn't it either.

I am slowly at my wit`s end and claim that I probably have a typo in there somewhere. The only question is: Where?

And since I haven't found any real guidance on this topic, I have a request: If anyone has the time and inclination to help me with the troubleshooting, here are the files: https://github.com/pftnhr/yellow-maillink/tree/unstable

Thank you in advance
Robert

**Comment by schulle4u** [comment, marked as answer]
What I can see at first glance is that your extension system and language settings are missing the correct formatting. Usually all extensions are using the format `ExtensionsnameSetting: value`. Assuming your extension has the name `Maillink`, a correctly formated extension or language setting would look like: 

```
    // Handle initialisation
    public function onLoad($yellow) {
        $this->yellow = $yellow;
        $this->yellow->system->setDefault("MaillinkAddress", "Insert your desired email address or remove this string");
        $this->yellow->language->setDefaults(array(
            "Language: it",
            "MaillinkLinktext: Lasciatemi un messaggio",
            "Language: de",
            "MaillinkLinktext: Schreib mir ein paar Zeilen",
            "Language: en",
            "MaillinkLinktext: Drop me a line",
            "Language: sv",
            "MaillinkLinktext: Skriv till mig",
            "Language: fr",
            "MaillinkLinktext: Contactez-moi"));
    }
```


I never tried another settings format, therefore don't know if this already solves your problem. At least your code will better fit into our coding standards if you are planning to publish your extension. BTW, please consider using spaces instead of tabs. 🙂

**Reply by pftnhr** [reply]
I followed your two pieces of advice and what can I say, it works!

Thank you so much Steffen!

**Comment by markseuffert** [comment]
The [API for developers](https://datenstrom.se/yellow/help/api-for-developers) provides these methods:

* `system->setDefault()` for system settings, each setting a pair with key and value
* `language->setDefaults()` for language settings, an array with multiple languages

Here's an example code:

```
    // Handle initialisation
    public function onLoad($yellow) {
        $this->yellow = $yellow;
        $this->yellow->system->setDefault("MaillinkEmail", "webmaster");
        $this->yellow->language->setDefaults(array(
            "Language: en",
            "MaillinkText: Drop me a line",
            "Language: de",
            "MaillinkText: Schreib mir ein paar Zeilen",
            "Language: sv",
            "MaillinkText: Hör av dig"));
    }
```

Here's an example code, this does NOT work:

```
    // Handle initialisation
    public function onLoad($yellow) {
        $this->yellow = $yellow;
        $this->yellow->system->setDefault("MailAddress", "");
        $this->yellow->system->setDefault("MailText", "Drop me a line");
    }
```

Here are some tips and further explanations. A key/value must not be empty, otherwise it will be ignored. Extensions should support the [languages of the official website](https://datenstrom.se/language/), which are `en de sv` at the time being. Extensions should use the same name in class/shortcuts/settings. That's how we avoid name collisions and it also makes it super easy to understand how different pieces fit together. Just by looking at the name.

Hope this helps and happy coding.

**Reply by pftnhr** [reply]
Hi Mark,

thank you very much for the detailed answer. Apparently it was due to the naming convention. I've bookmarked your comment so I can check it the next time something goes wrong.
